File: controllers/store.js
const jwt = require('jsonwebtoken');

const Article = require('../models/article');

const auth = require('../utils/auth');

// afficher un article
exports.showOne = (req, res, next) => {
    try {
        if (!req.body.token) {
            throw "NoToken"
        }
        if (!/[0-9a-f]{12}/g.test(req.params.id)) {
            throw "BadIdFormat"
        }
        const token = req.body.token;
        const id = req.params.id
        auth.isAdmin(token)
            .then(admin => {
                if (admin) {
                    Article.findOne({ _id: id })
                        .then(article => {
                            if (article) {
                                res.status(200).json(article)
                            } else {
                                res.sendStatus(404)
                            }
                        })
                        .catch(error => {
                            console.error(error);
                            res.sendStatus(500);
                        });
                } else {
                    res.sendStatus(403);
                }
            })
            .catch(error => {
                if (error instanceof jwt.TokenExpiredError) {
                    res.status(401).json({ error });
                } else if (error instanceof jwt.JsonWebTokenError) {
                    res.status(401).json({ error });
                } else {
                    console.error(error);
                    res.sendStatus(500);
                }
            });
    }
    catch (error) {
        if (error == "NoToken") {
            res.status(401).json({ error });
        } else if (error == "BadIdFormat") {
            res.status(400).json({ error });
        } else {
            res.sendStatus(500);
        }

    }
};

// creer un article
exports.create = (req, res, next) => {
    try {
        if (!req.body.token) {
            throw "NoToken"
        }
        if (!req.body.name) {
            throw "NoName"
        }
        const token = req.body.token;
        const name = req.body.name

        delete req.body.token;
        auth.isAdmin(token)
            .then(admin => {
                if (admin) {
                    Article.findOne({ name: name })
                        .then(article => {
                            if (!article) {
                                const article = new Article({ ...req.body });
                                article.save()
                                    .then(() => {
                                        Article.findOne({ name: name })
                                            .then(article => res.status(201).json({ id: article._id }))
                                            .catch(error => {
                                                console.error(error);
                                                res.sendStatus(500);
                                            });
                                    })
                                    .catch(error => {
                                        console.error(error);
                                        res.sendStatus(500);
                                    });
                            } else {
                                res.status(400).json({ id: article._id })
                            }
                        })
                        .catch(error => {
                            console.error(error);
                            res.sendStatus(500);
                        });
                } else {
                    res.sendStatus(403);
                }
            })
            .catch(error => {
                if (error instanceof jwt.TokenExpiredError) {
                    res.status(401).json({ error });
                } else if (error instanceof jwt.JsonWebTokenError) {
                    res.status(401).json({ error });
                } else {
                    console.error(error);
                    res.sendStatus(500);
                }
            });
    }
    catch (error) {
        if (error == "NoToken") {
            res.status(401).json({ error });
        } else if (error == "NoName") {
            res.status(400).json({ error });
        } else if (error == "BadIdFormat") {
            res.status(400).json({ error });
        } else {
            res.sendStatus(500);
        }

    }
};

//modifier un article
exports.modify = (req, res, next) => {
    try {
        if (!req.body.token) {
            throw "NoToken"
        }
        if (!/[0-9a-f]{12}/g.test(req.params.id)) {
            throw "BadIdFormat"
        }
        const token = req.body.token;
        const id = req.params.id
        delete req.body.token;
        auth.isAdmin(token)
            .then(admin => {
                if (admin) {

                    Article.findOne({ _id: id })
                        .then(article => {
                            if (article) {
                                Article.updateOne({ _id: id }, { ...req.body, _id: id })
                                    .then(article => res.sendStatus(200))
                                    .catch(error => {
                                        console.error(error);
                                        res.sendStatus(500);
                                    });
                            } else {
                                res.sendStatus(404)
                            }
                        })
                        .catch(error => {
                            console.error(error);
                            res.sendStatus(500);
                        });

                } else {
                    res.sendStatus(403)
                }
            })
            .catch(error => {
                if (error instanceof jwt.TokenExpiredError) {
                    res.status(401).json({ error });
                } else if (error instanceof jwt.JsonWebTokenError) {
                    res.status(401).json({ error });
                } else {
                    console.error(error);
                    res.sendStatus(500);
                }
            });
    }
    catch (error) {
        if (error == "NoToken") {
            res.status(401).json({ error });
        } else if(error == "BadIdFormat"){
            res.status(400).json({ error });
        } else {
            res.sendStatus(500);
        }

    }
};

//supprimer un article
exports.del = (req, res, next) => {
    try {
        if (!req.body.token) {
            throw "NoToken"
        }
        if (!/[0-9a-f]{12}/g.test(req.params.id)) {
            throw "BadIdFormat"
        }
        const token = req.body.token;
        const id = req.params.id
        auth.isAdmin(token)
            .then(admin => {
                if (admin) {

                    Article.findOne({ _id: id })
                        .then(article => {
                            if (article) {
                                Article.deleteOne({ _id: id })
                                    .then(() => res.sendStatus(200))
                                    .catch(error => {
                                        console.error(error);
                                        res.sendStatus(500);
                                    });
                            } else {
                                res.sendStatus(404)
                            }
                        })
                        .catch(error => {
                            console.error(error);
                            res.sendStatus(500);
                        });
                } else {
                    res.sendStatus(403);
                }
            })
            .catch(error => {
                if (error instanceof jwt.TokenExpiredError) {
                    res.status(401).json({ error });
                } else if (error instanceof jwt.JsonWebTokenError) {
                    res.status(401).json({ error });
                } else {
                    console.error(error);
                    res.sendStatus(500);
                }
            });
    }
    catch (error) {
        if (error == "NoToken") {
            res.status(401).json({ error });
        } else if (error == "BadIdFormat") {
            res.status(400).json({ error });
        } else {
            res.sendStatus(500);
        }

    }
};

// afficher tout le magasin
exports.showAll = (req, res, next) => {
    Article.find()
        .then(articles => res.status(200).json(articles))
        .catch(error => res.status(500).json({ error }));
};

File: .client/store.lua
local http = require 'socket.http'
local mime = require 'mime'
local ltn12 = require 'ltn12'
local json = require 'json'

local store = {}

-- afficher un article
function store.showOne(url, tokenSess, idToShow)
    local urlString = url.."/store/"..idToShow
    local req = json.encode({token = tokenSess})
    local res = {}
    local name = 'nil'
    local prix = 'nil'
    local body = {}

    local result, statuscode, headers, statustext = http.request {
        method = "GET",
        url = urlString,
        source = ltn12.source.string(req),
        headers={
            ["content-type"] = "application/json; charset=utf-8",
            ["content-length"] = tostring(#req)
        },
        sink = ltn12.sink.table(res)
    }

    if statuscode == 200 then
        body = json.decode(table.concat(res))
        name = body.name
        prix = tonumber(body.prix["$numberDecimal"],10)
        end
    
        return statuscode, name, prix
end

-- créer un article
function store.create(url, tokenSess, artName, descript, artPrix)
    local urlString = url.."/store/"
    local createString = {token = tokenSess, name = artName, desc = descript}
    if artPrix then
        createString.prix = artPrix;
    end
    local req = json.encode(createString)
    local res = {}
    local artId = 'nil'
    local body = {}

    local result, statuscode, headers, statustext = http.request {
        method = "POST",
        url = urlString,
        source = ltn12.source.string(req),
        headers={
            ["content-type"] = "application/json; charset=utf-8",
            ["content-length"] = tostring(#req)
        },
        sink = ltn12.sink.table(res)
    }

    
    if statuscode == 201 or statuscode == 400 then
        body = json.decode(table.concat(res))
        artId = body.message
    end
    return statuscode, artId

end

-- modifier un article
function store.modify(url, tokenSess, idToChange, newName, newDesc, newPrix)
    local modifyString = {token = tokenSess};
    if newName then
        modifyString.name = newName;
    end
    if newDesc then
        modifyString.desc = newDesc;
    end
    if newRole then
        modifyString.prix = newPrix;
    end

    local urlString = url.."/store/"..idToChange
    local req = json.encode(modifyString)
    local res = {}

    local result, statuscode, headers, statustext = http.request {
        method = "PUT",
        url = urlString,
        source = ltn12.source.string(req),
        headers={
            ["content-type"] = "application/json; charset=utf-8",
            ["content-length"] = tostring(#req)
        },
        sink = ltn12.sink.table(res)
    }

    return statuscode

end

-- supprimer un article
function store.del(url, tokenSess, idToDel)

    local urlString = url.."/store/"..idToDel
    local req = json.encode({token = tokenSess})
    local res = {}

    local result, statuscode, headers, statustext = http.request {
        method = "DELETE",
        url = urlString,
        source = ltn12.source.string(req),
        headers={
            ["content-type"] = "application/json; charset=utf-8",
            ["content-length"] = tostring(#req)
        },
        sink = ltn12.sink.table(res)
    }

    return statuscode
end

-- afficher tout le magasin
function store.showAll(url)
    local urlString = url.."/store/"
    local res = {};
    local body = {};

    local result, statuscode, headers, statustext = http.request {
        method = "GET",
        url = urlString,
        headers={
            ["content-type"] = "application/json; charset=utf-8",
        },
        sink = ltn12.sink.table(res)
    }

    if statuscode == 200 then
        body = json.decode(table.concat(res));
    end

    return statuscode, body;
end

return store